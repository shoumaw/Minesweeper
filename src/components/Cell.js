import React, { useState, useEffect, useRef, forwardRef, useImperativeHandle } from 'react';
import {
    StyleSheet,
    View,
    TouchableOpacity,
    Image,
    Text
} from 'react-native';
import Constants from '../utils/AppConstants';
import Images from '../../assets/Images';
export default Cell = forwardRef(({ onUncover: onUncoverCb, setTotalNumUrchins, onGameOver, width, height, x, y }, ref) => {
    const [isUncovered, changeVisibility] = useState({ visibility: false, user: false })
    const [isUrchin, setUrchinProbability] = useState(Math.random() < 0.1)
    const [urchinsAround, setUrchinsAround] = useState(null)
    const isInitialMount = useRef(true);
    // Incerement the total number of urchins on the board
    if (isUrchin) {
        setTotalNumUrchins()
    }
    // OnUncover side effect. Its either gameover or call the parent uncover callback
    useEffect(() => {
        if (isInitialMount.current) {
            isInitialMount.current = false;
        } else if (isUncovered.user) {
            if (isUrchin) {
                onGameOver(Constants.YOU_LOSE);
            } else {
                onUncoverCb(x, y);
            }
        }

    }, [isUncovered])
    // On uncover change visibility 
    const onUncover = (userInitiated) => {

        if (isUncovered.visibility) {
            return;
        }

        if (!userInitiated && isUrchin) {
            return;
        }
        changeVisibility({ visibility: true, user: true })

    }
    // Uncover all cells when the game is over
    const uncoverWithoutSideEffects = () => {
        if (isUncovered.visibility) {
            return;
        }

        changeVisibility({ visibility: true, user: false })

    }
    // Ref functions to be called using the parent component
    useImperativeHandle(ref, () => ({
        onUncover,
        uncoverWithoutSideEffects,
        reset: () => {
            changeVisibility({ visibility: false, user: false })
            setUrchinProbability(Math.random() < 0.2)
            setUrchinsAround(null)
        },
        isUrchin: () => {
            return isUrchin
        },
        setUrchinsAround: (urchinsAround) => {
            setUrchinsAround(urchinsAround)
        }
    }));


    const cell = () => {
        if (!isUncovered.visibility) {
            return (
                <TouchableOpacity onPress={() => onUncover(true)}>
                    <View style={[styles.cell, { width: width, height: height }]}>

                    </View>
                </TouchableOpacity>
            )
        } else {
            let content = null;
            if (isUrchin) {

                content = <Image id="urchin-image" source={Images.urchin} style={{ width: width / 2, height: height / 2 }} resizeMode="contain" />

            } else if (urchinsAround) {
                content = <Text id="num-urchins">{urchinsAround}</Text>

            }
            return (<View id="cell-view" style={[styles.cellUncovered, { width: width, height: height }]}>
                {content}
            </View>)
        }
    }

    return (cell())


})

const styles = StyleSheet.create({
    cell: {
        backgroundColor: '#bdbdbd',
        borderWidth: 3,
        borderTopColor: '#ffffff',
        borderLeftColor: '#ffffff',
        borderRightColor: '#7d7d7d',
        borderBottomColor: '#7d7d7d'
    },
    cellUncovered: {
        backgroundColor: '#ffffff',
        borderWidth: 1,
        borderColor: '#7d7d7d',
        alignItems: 'center',
        justifyContent: 'center'
    }

})