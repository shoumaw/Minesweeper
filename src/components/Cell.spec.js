import React from 'react';
import Cell from './Cell';
import Constants from '../utils/AppConstants'
import renderer from 'react-test-renderer';
import { shallow, configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import {
    View,
    Image,
    Text
} from 'react-native';
configure({ adapter: new Adapter() })
test('renders correctly', () => {
    const tree = renderer.create(
        <Cell
            onUncover={jest.fn()}
            onGameOver={jest.fn()}
            setTotalNumUrchins={jest.fn()}
            key={0}
            width={Constants.CELL_SIZE}
            height={Constants.CELL_SIZE}
            x={0}
            y={0}
            ref={null}
        />).toJSON();
    expect(tree).toMatchSnapshot();
});

test('renders children correctly', () => {
    const wrapper = shallow(
        <Cell
            onUncover={jest.fn()}
            onGameOver={jest.fn()}
            setTotalNumUrchins={jest.fn()}
            key={0}
            width={Constants.CELL_SIZE}
            height={Constants.CELL_SIZE}
            x={0}
            y={0}
            ref={null} />
    );
    expect(wrapper.find(View).exists()).toBeTruthy()
    expect(wrapper.find(Text).exists()).toBeFalsy()
    expect(wrapper.find(Image).exists()).toBeFalsy()

});


