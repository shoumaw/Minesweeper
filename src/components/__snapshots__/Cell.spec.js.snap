// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#bdbdbd",
          "borderBottomColor": "#7d7d7d",
          "borderLeftColor": "#ffffff",
          "borderRightColor": "#7d7d7d",
          "borderTopColor": "#ffffff",
          "borderWidth": 3,
        },
        Object {
          "height": 30,
          "width": 30,
        },
      ]
    }
  />
</View>
`;
